controllers/awscluster_controller_test.go

```diff
@@ -13,42 +13,1154 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+func TestAWSClusterReconciler_IntegrationTests(t *testing.T) {
+	var (
+		reconciler AWSClusterReconciler
+		mockCtrl   *gomock.Controller
+		recorder   *record.FakeRecorder
+	)
+	var secret = &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-secret",
+			Namespace: "capa-system",
+		},
+		Data: map[string][]byte{
+			"AccessKeyID":     []byte("access-key-id"),
+			"SecretAccessKey": []byte("secret-access-key"),
+			"SessionToken":    []byte("session-token"),
+		},
+	}
+
+	setup := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		recorder = record.NewFakeRecorder(10)
+		reconciler = AWSClusterReconciler{
+			Client:   testEnv.Client,
+			Recorder: recorder,
+		}
+	}
 
-	reconciler := &AWSClusterReconciler{
-		Client: testEnv.Client,
+	teardown := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl.Finish()
 	}
 
-	instance := &infrav1.AWSCluster{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"}}
-	instance.Default()
+	t.Run("Should not create VPC in case of unmanaged VPC", func(t *testing.T) {
+		g := NewWithT(t)
+		mockCtrl = gomock.NewController(t)
+		ec2Mock := mock_ec2iface.NewMockEC2API(mockCtrl)
+		elbMock := mock_elbiface.NewMockELBAPI(mockCtrl)
+		expect := func(m *mock_ec2iface.MockEC2APIMockRecorder, e *mock_elbiface.MockELBAPIMockRecorder) {
+			mockedCreateVPCCalls(m)
+			mockedCreateSGCalls(m)
+			mockedCreateLBCalls(e)
+			mockedDescribeInstanceCall(m)
+		}
+		expect(ec2Mock.EXPECT(), elbMock.EXPECT())
+
+		capaNS := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "capa-system",
+			},
+		}
+		g.Expect(testEnv.Create(ctx, capaNS)).To(Succeed())
+
+		controllerIdentity := createControllerIdentity(g)
+		ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("integ-test-%s", util.RandomString(5)))
+		g.Expect(err).To(BeNil())
+		setup(t, g)
+		awsCluster := getAWSCluster("test", ns.Name)
+		g.Expect(reconciler.Client.Create(ctx, secret)).To(Succeed())
+
+		g.Expect(testEnv.Create(ctx, &awsCluster)).To(Succeed())
+		defer teardown(t, g)
+		defer t.Cleanup(func() {
+			g.Expect(testEnv.Cleanup(ctx, controllerIdentity)).To(Succeed())
+			g.Expect(testEnv.Cleanup(ctx, &awsCluster)).To(Succeed())
+		})
+
+		cs, err := scope.NewClusterScope(
+			scope.ClusterScopeParams{
+				Client: testEnv.Client,
+				Cluster: &clusterv1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-cluster",
+					},
+				},
+				AWSCluster: &awsCluster,
+			},
+		)
+		g.Expect(err).To(BeNil())
+		networkSvc := network.NewService(cs)
+		networkSvc.EC2Client = ec2Mock
+		reconciler.networkServiceFactory = func(clusterScope scope.ClusterScope) services.NetworkInterface {
+			return networkSvc
+		}
+
+		ec2Svc := ec2Service.NewService(cs)
+		ec2Svc.EC2Client = ec2Mock
+		reconciler.ec2ServiceFactory = func(scope scope.EC2Scope) services.EC2MachineInterface {
+			return ec2Svc
+		}
+		_, err = ec2Svc.EC2Client.CreateSubnet(&ec2.CreateSubnetInput{
```

Comment:
What is the purpose of this CreateSubnet() call here?

Reply:
In case of unmanaged VPC, we should have subnet pre created, see the doc [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/book/src/topics/consuming-existing-aws-infrastructure.md#consuming-existing-aws-infrastructure)

Reply:
It is required to precreate these resources when calling actual AWS apis like in case of e2e tests, but here in integration tests we are mocking the AWS calls so we don’t need to add these calls to satisfy AWS related prerequisites. This statement is not doing anything during this test so we can remove it.
Similar concern is there in other integration test - ‘Should fail VPC creation if VPC limit exceeded’ 
`_, err = s.EC2Client.CreateVpc(&ec2.CreateVpcInput{})`
This statement is also not required, we can return error in the mocked CreateVpc() call without making these dummy calls in test.

Reply:
If I do not pre-create the subnet here, it gives error and gets stuck at this [point](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/323f0a8506ca9483006e381cc9ac26fb8743dcbe/pkg/cloud/services/ec2/bastion.go#L62-L68). i have added it only to overcome this issue.

Reply:
> If I do not pre-create the subnet here, it gives error and gets stuck at this point.

Tried it on my system, it is working fine.

Reply:
I will give it a try.

Reply:
seems right, thanks :) maybe with some mocking that changed that I didn't notice. I will remove it.